### PR TITLE
Node AutoLayout

### PR DESCRIPTION
**Pull request by QilongTang**
Please Note:
1. Before submitting the PR, please review [How to Contribute to Dynamo](https://github.com/DynamoDS/Dynamo/blob/master/CONTRIBUTING.md)
2. Dynamo Team will meet 1x a month to review PRs found on Github (Issues will be handled separately)
3. PRs will be reviewed from oldest to newest
4. If a reviewed PR requires changes by the owner, the owner of the PR has 30 days to respond. If the PR has seen no activity by the next session, it will be either closed by the team or depending on its utility will be taken over by someone on the team
5. PRs should use either Dynamo's default PR template or [one of these other template options](https://github.com/DynamoDS/Dynamo/wiki/Choosing-a-Pull-Request-Template) in order to be considered for review.
6. PRs that do not have one of the Dynamo PR templates completely filled out with all declarations satisfied will not be reviewed by the Dynamo team.
7. PRs made to the `DynamoRevit` repo will need to be cherry-picked into all the DynamoRevit Release branches that Dynamo supports. Contributors will be responsible for cherry-picking their reviewed commits to the other branches after a `LGTM` label is added to the PR.

### Purpose

Auto layout the newly created node from node auto complete.

Instead of the existing auto layout algorithm we are using for clean up layout which makes the initial node jumpy, I am simply putting the new node right next to the input port and arrange some Y offset based on inputport index. 

AutoLayout_v3

### Declarations

Check these if you believe they are true

- [x] The codebase is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [x] All tests pass using the self-service CI.
- [x] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated

### Reviewers

@DynamoDS/dynamo 

### FYIs

(FILL ME IN, Optional) Names of anyone else you wish to be notified of

### CODE REVIEW

File: src/DynamoCoreWpf/ViewModels/Search/SearchViewModel.cs [hunk 1]
```diff
@@ -1061,11 +1062,20 @@ public void OnSearchElementClicked(NodeModel nodeModel, Point position)
...
         {
+            // Do not auto connect code block node since default code block node do not have output port
             if (!nodeCreationName.Contains("Code Block"))
             {
+                //dynamoViewModel.ExecuteCommand(new DynamoModel.MakeConnectionCommand(nodeModel.GUID,))
+                var id = Guid.NewGuid();
                 // Create a new node based on node creation name and connect ports
-                dynamoViewModel.ExecuteCommand(new DynamoModel.CreateAndConnectNodeCommand(Guid.NewGuid(), portModel.Owner.GUID,
-                    nodeCreationName, 0, portModel.Index, portModel.CenterX - 50, portModel.CenterY + 200, false, false));
+                var initialNode = dynamoViewModel.CurrentSpaceViewModel.Nodes.Where(x => x.Id == portModel.Owner.GUID).FirstOrDefault();
```
Comment (aparajit-pratap) [inline src/DynamoCoreWpf/ViewModels/Search/SearchViewModel.cs:1071]: isn't `initialNode` the same as `portModel.Owner`?

Comment (aparajit-pratap) [reply src/DynamoCoreWpf/ViewModels/Search/SearchViewModel.cs:1071]: Can we split this step into a `CreateNodeCommand` step and then add a new command called something like `PlaceAndConnectCommand`? That way we can first create the new node, get its height and width and place it relative to the original node according to its respective height and width. In other words the `PlaceAndConnectCommand` will be a combination of 2 steps: 1) place the node based on an calculated X and Y position and 2) call the `MakeConnectionCommand`.

Comment (aparajit-pratap) [reply src/DynamoCoreWpf/ViewModels/Search/SearchViewModel.cs:1071]: This way we might be able to do away with these hard-coded values.

File: src/DynamoCoreWpf/ViewModels/Search/SearchViewModel.cs [hunk 2]
```diff
@@ -1061,11 +1062,20 @@ public void OnSearchElementClicked(NodeModel nodeModel, Point position)
...
+                // Placing the new node based on which input port it is connecting to.
+                // The larger index for the input port, the lower the new node will be placed. 60 and 200 and offset we can adjust.
+                double adjustedY = initialNode.Y + (portModel.Index - 1) * 60 ;
+                // Placing the new node to the left of initial node
+                var adjustedX = initialNode.X - 200; 
+                dynamoViewModel.ExecuteCommand(new DynamoModel.CreateAndConnectNodeCommand(id, portModel.Owner.GUID,
+                    nodeCreationName, 0, portModel.Index, adjustedX, adjustedY, false, false));
```
Comment (aparajit-pratap) [inline src/DynamoCoreWpf/ViewModels/Search/SearchViewModel.cs:1078]: Can we split this step into a CreateNodeCommand step and then add a new command called something like ``PlaceAndConnectCommand``? That way we can first create the new node, get its height and width and use the height and width to place it relative to the original node. In other words, the `PlaceAndConnectCommand`
will be a combination of 2 steps: 1) place the node based on a calculated X and Y position, and 2) call the
`MakeConnectionCommand`. This way we might be able to do away with these hard-coded values.

Comment (QilongTang) [reply src/DynamoCoreWpf/ViewModels/Search/SearchViewModel.cs:1078]: @aparajit-pratap I did thought about that before implementation but the height and width for the new nodes are always default to 100 for node model..

Comment (QilongTang) [reply src/DynamoCoreWpf/ViewModels/Search/SearchViewModel.cs:1078]: image
@aparajit-pratap This is what I am referring to. I already considered this solution before finalizing the PR. For now there isn't a way I can see we can get the actual height and width for the newly placed node.

Comment (QilongTang) [reply src/DynamoCoreWpf/ViewModels/Search/SearchViewModel.cs:1078]: Another slightly less important reason is about undo/redo, the current implementation only requires one undo to get back to the previous state versus the split way will need to be done by two undos.

Comment (aparajit-pratap) [reply src/DynamoCoreWpf/ViewModels/Search/SearchViewModel.cs:1078]: I'm trying something else, will get back to you on this PR, I may commit directly to this PR if I can get something working.

Comment (aparajit-pratap) [reply src/DynamoCoreWpf/ViewModels/Search/SearchViewModel.cs:1078]: Couldn't get it to work. In any case this is the problem with this approach right now:
create_and_connect_nodes

Comment (QilongTang) [reply src/DynamoCoreWpf/ViewModels/Search/SearchViewModel.cs:1078]: Indeed, placing any slider or any giant node will result in such problem but I dont see a way to be smarter yet..

**Review by aparajit-pratap — COMMENTED** [review]
@QilongTang I think we can close this :)

**Pull request closed without merging**